# 4.9. Answers - Part 1

## This script contains:
#### 01. Importing Libraries
#### 02. Set Path
#### 03. Importing data
#### 04. CUSTOMERS DATA
##### 4.1. Checking data
##### 4.2. Wrangling data
###### 4.2.1. Renaming Columns
###### 4.2.2. Changing a Variable's Data Type
###### 4.2.3. Checking for missing values
###### 4.2.4. Checking for duplicates
###### 4.2.5. Checking for mixed-type data
##### 4.3. Last check-up 
#### 05. ORDERS_PRODUCTS_MERGED_AGGREGATED_CHARTS DATA
#### 06. Merging Data
#### 07. Checking merged data
#### 08. Export Data

## 01. Importing Libraries

In [1]:
# Importing Data

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

## 02. Set Path

In [2]:
# Set path

path = r'C:\Users\Asus\OneDrive\Documents\Data Analytics\Data Immersion\4. Python Fundamentals for Data Analysts\05-05-2023 Instacart Basket Analysis'

In [3]:
# Check path

path

'C:\\Users\\Asus\\OneDrive\\Documents\\Data Analytics\\Data Immersion\\4. Python Fundamentals for Data Analysts\\05-05-2023 Instacart Basket Analysis'

## 03. Importing data

In [4]:
# Import new data - customers.csv

df_custom = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [5]:
# Import old data - orders_products_merge_agg_charts.pkl

df_ords_prods_merge_agg_charts = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merge_agg_charts.pkl'))

# 04. CUSTOMERS DATA

## 4.1. Checking data

In [6]:
# Check customers.csv

df_custom.head(30)

user_id First Name     Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah   Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia       Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth     Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle      Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann    Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia      Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris     Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph    Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy       Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn      Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria   Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger      Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia   Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie     Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter       Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon       Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy       Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian     Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor       Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick     Fowler    Male         Vermont   65    1/1/2017   
20    29864      Randy      Mills    Male    North Dakota   39    1/1/2017   
21    37744      Wanda      Salas  Female        Arkansas   77    1/1/2017   
22    50000     Sandra      Trejo  Female           Maine   64    1/1/2017   
23    55567     Martha      Hurst  Female  North Carolina   67    1/1/2017   
24    57549      Shawn      Moore    Male         Montana   72    1/1/2017   
25    60386      Wayne   Woodward    Male          Alaska   59    1/1/2017   
26    68032    Charles  Stevenson    Male   West Virginia   74    1/1/2017   
27    83910       Ruth       Rice  Female         Indiana   70    1/1/2017   
28    86936   Kimberly     Monroe  Female          Oregon   57    1/1/2017   
29    99199   Kimberly     Monroe  Female        Arkansas   80    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559  
15             0                  divorced/widowed  100091  
16             3                           married   57096  
17             1                           married   91826  
18             2                           married   91111  
19             1                           married   34514  
20             1                           married   44364  
21             1                           married  125977  
22             2                           married  115489  
23             2                           married   48645  
24             0                  divorced/widowed  135302  

In [7]:
# Dimension of customers

df_custom.shape

(206209, 10)

In [8]:
# Information about customers dataframe

df_custom.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [9]:
# Descriptive statistics of customers

df_custom.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

## 4.2. Wrangling Data

### 4.2.1. Renaming Columns

In [10]:
# Renaming column 'Surnam' to 'Last Name'

df_custom.rename(columns = {'Surnam' : 'Last_Name'}, inplace = True)

In [11]:
# Renaming column 'n_dependants' to 'Number_of_dependants'

df_custom.rename(columns = {'n_dependants' : 'Number_of_dependants'}, inplace = True)

In [12]:
# Renaming column 'fam_status' to 'Family_status'

df_custom.rename(columns = {'fam_status' : 'Family_status'}, inplace = True)

In [13]:
# Renaming column 'First Name' to 'First_Name'

df_custom.rename(columns = {'First Name' : 'First_Name'}, inplace = True)

In [14]:
# Renaming column 'user_id' to 'User_ID'

df_custom.rename(columns = {'user_id' : 'User_ID'}, inplace = True)

In [15]:
# Renaming column 'date_joined' to 'Date_joined'

df_custom.rename(columns = {'date_joined' : 'Date_joined'}, inplace = True)

In [16]:
# Renaming column 'income' to 'Income'

df_custom.rename(columns = {'income' : 'Income'}, inplace = True)

In [17]:
# Checking renamed columns

df_custom.head(10)

User_ID First_Name Last_Name  Gender           STATE  Age Date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   Number_of_dependants Family_status  Income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374  
5                     2       married   49643  
6                     0        single   61746  
7                     0        single   63712  
8                     1       married  162432  
9                     2       married   32072

### 4.2.2. Changing a Variable's Data Type

In [18]:
# Changing the data type of the 'User_ID' column from a number to a string

df_custom['User_ID'] = df_custom['User_ID'].astype('int32')
df_custom['Age'] = df_custom['Age'].astype('int32')

In [19]:
# Checking data type

df_custom['User_ID'].dtype

dtype('int32')

### 4.2.3. Checking for missing values

In [20]:
# Checking for missing values

df_custom.isnull().sum()

User_ID                     0
First_Name              11259
Last_Name                   0
Gender                      0
STATE                       0
Age                         0
Date_joined                 0
Number_of_dependants        0
Family_status               0
Income                      0
dtype: int64

In [21]:
# Creating a subset of the dataframe containing only the values in question

df_nan = df_custom[df_custom['First_Name'].isnull() == True]

In [22]:
# Checking df_nan

df_nan

User_ID First_Name Last_Name  Gender         STATE  Age Date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        Number_of_dependants Family_status  Income  
53                         2       married   41709  
73                         0        single   82518  
82                         3       married  117099  
99                         1       married  155673  
105                        3       married  151819  
...                      ...           ...     ...  
206038                     3       married   87783  
206044                     2       married  108488  
206090                     1       married   45275  
206162                     0        single   56325  
206171                     2       married   59222  

[11259 rows x 10 columns]

#### Although we have some missing first names (11259 NaN) out of 206209 rows, we will not do anything about it as this column is not relevant for our analysis, but it is still meaningful to us as other information in the same rows are important to us.

In [23]:
# Changing the data type of the 'First_Name' column from a 'object' to a 'string'

df_custom['First_Name'] = df_custom['First_Name'].astype('str')

In [24]:
# Checking for missing values in First_Name column

df_custom['First_Name'].isnull().sum()

0

#### After changing data type of "First_Name" column from 'object' to a 'str' there are no more any missing values (NaN's).

### 4.2.4. Checking for duplicates

In [25]:
# Looking for full duplicates within our dataframe

df_dups = df_custom[df_custom.duplicated()]

In [26]:
# Checking if there are any duplicates

df_dups

Empty DataFrame
Columns: [User_ID, First_Name, Last_Name, Gender, STATE, Age, Date_joined, Number_of_dependants, Family_status, Income]
Index: []

#### There are no duplicates in customers dataframe.

### 4.2.5. Checking for mixed-type data

In [27]:
# Check for mixed types

for col in df_custom.columns.tolist():
    weird = (df_custom[[col]].applymap(type) != df_custom[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_custom[weird]) > 0:
        print (col)
    else:
        print ('No mixed dtypes', col)
    


No mixed dtypes User_ID
No mixed dtypes First_Name
No mixed dtypes Last_Name
No mixed dtypes Gender
No mixed dtypes STATE
No mixed dtypes Age
No mixed dtypes Date_joined
No mixed dtypes Number_of_dependants
No mixed dtypes Family_status
No mixed dtypes Income


#### No mixed-type data in columns dataframe.

## 4.3. Last check-up

In [28]:
# Check customers

df_custom.head(15)

User_ID First_Name Last_Name  Gender           STATE  Age Date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   

    Number_of_dependants                     Family_status  Income  
0                      3                           married  165665  
1                      0                            single   59285  
2                      2                           married   99568  
3                      0                            single   42049  
4                      1                           married   40374  
5                      2                           married   49643  
6                      0                            single   61746  
7                      0                            single   63712  
8                      1                           married  162432  
9                      2                           married   32072  
10                     2  living with parents and siblings   43443  
11                     2                           married   47988  
12                     1                           married   57835  
13                     1                           married   71190  
14                     2                           married  146559

In [29]:
# Information about customers dataframe

df_custom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   User_ID               206209 non-null  int32 
 1   First_Name            206209 non-null  object
 2   Last_Name             206209 non-null  object
 3   Gender                206209 non-null  object
 4   STATE                 206209 non-null  object
 5   Age                   206209 non-null  int32 
 6   Date_joined           206209 non-null  object
 7   Number_of_dependants  206209 non-null  int64 
 8   Family_status         206209 non-null  object
 9   Income                206209 non-null  int64 
dtypes: int32(2), int64(2), object(6)
memory usage: 14.2+ MB


In [30]:
# Checking for missing values

df_custom.isnull().sum()

User_ID                 0
First_Name              0
Last_Name               0
Gender                  0
STATE                   0
Age                     0
Date_joined             0
Number_of_dependants    0
Family_status           0
Income                  0
dtype: int64

In [31]:
# Descriptive statistics of customers

df_custom.describe()

User_ID            Age  Number_of_dependants         Income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

In [32]:
# Dimension of customers

df_custom.shape

(206209, 10)

# 05. ORDERS_PRODUCTS_MERGED_AGGREGATED_CHARTS DATA

In [33]:
# Check customers.csv

df_ords_prods_merge_agg_charts.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   
5   1701441      777            16                   1                  7   
6   1871483      825             3                   2                 14   
7   1290456      910            12                   3                 10   
8    369558     1052            10                   1                 20   
9    589712     1052            15                   1                 12   

   ...    price_range_loc     busiest_day    busiest_days  \
0  ...  Mid-range product  Regularly busy  Regularly busy   
1  ...  Mid-range product  Regularly busy  Regularly busy   
2  ...  Mid-range product    Bussiest day    Busiest days   
3  ...  Mid-range product  Regularly busy    Slowest days   
4  ...  Mid-range product      Least busy    Slowest days   
5  ...  Mid-range product  Regularly busy    Busiest days   
6  ...  Mid-range product  Regularly busy  Regularly busy   
7  ...  Mid-range product  Regularly busy    Slowest days   
8  ...  Mid-range product  Regularly busy    Busiest days   
9  ...  Mid-range product  Regularly busy    Busiest days   

  busiest_period_of_day max_order      loyalty_flag avg_price spending_flag  \
0           Most orders        32  Regular customer  6.935811   Low spender   
1        Average orders        32  Regular customer  6.935811   Low spender   
2        Average orders         5      New customer  7.930208   Low spender   
3           Most orders         3      New customer  4.972414   Low spender   
4        Average orders         3      New customer  4.972414   Low spender   
5        Average orders        26  Regular customer  6.935398   Low spender   
6           Most orders         9      New customer  5.957576   Low spender   
7           Most orders        12  Regular customer  6.680000   Low spender   
8        Average orders        20  Regular customer  7.162500   Low spender   
9           Most orders        20  Regular customer  7.162500   Low spender   

  median_frequency  order_frequency_flag  
0              8.0     Frequent customer  
1              8.0     Frequent customer  
2              8.0     Frequent customer  
3              9.0     Frequent customer  
4              9.0     Frequent customer  
5             11.0      Regular customer  
6             20.0      Regular customer  
7              6.0     Frequent customer  
8             10.0     Frequent customer  
9             10.0     Frequent customer  

[10 rows x 25 columns]

In [34]:
# Renaming column 'user_id' to 'User_ID'

df_ords_prods_merge_agg_charts.rename(columns = {'user_id' : 'User_ID'}, inplace = True)

In [35]:
# Dimension of df_ords_prods_merge_agg_charts

df_ords_prods_merge_agg_charts.shape

(32404859, 25)

In [36]:
# Information about df_ords_prods_merge_agg_charts dataframe

df_ords_prods_merge_agg_charts.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   product_id              int32   
 1   product_name            object  
 2   aisle_id                int32   
 3   department_id           int32   
 4   prices                  float64 
 5   order_id                int32   
 6   User_ID                 int32   
 7   order_number            int8    
 8   orders_day_of_week      int8    
 9   order_hour_of_day       int8    
 10  days_since_prior_order  float16 
 11  add_to_cart_order       float16 
 12  reordered               float16 
 13  _merge                  category
 14  merge_products          category
 15  price_range_loc         object  
 16  busiest_day             object  
 17  busiest_days            object  
 18  busiest_period_of_day   object  
 19  max_order               int8    
 20  loyalty_flag            object  
 21  avg_pr

In [37]:
# Descriptive statistics of df_ords_prods_merge_agg_charts

df_ords_prods_merge_agg_charts.describe()

product_id      aisle_id  department_id        prices      order_id  \
count  3.240486e+07  3.240486e+07   3.240486e+07  3.239973e+07  3.240486e+07   
mean   2.559866e+04  7.119612e+01   9.919792e+00  7.790994e+00  1.710745e+06   
std    1.408400e+04  3.821139e+01   6.281485e+00  4.241809e+00  9.872988e+05   
min    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  2.000000e+00   
25%    1.354400e+04  3.100000e+01   4.000000e+00  4.200000e+00  8.559470e+05   
50%    2.530200e+04  8.300000e+01   9.000000e+00  7.400000e+00  1.711049e+06   
75%    3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01  2.565499e+06   
max    4.968800e+04  1.340000e+02   2.100000e+01  2.500000e+01  3.421083e+06   

            User_ID  order_number  orders_day_of_week  order_hour_of_day  \
count  3.240486e+07  3.240486e+07        3.240486e+07       3.240486e+07   
mean   1.029372e+05  1.714230e+01        2.738867e+00       1.342515e+01   
std    5.946610e+04  1.753532e+01        2.090077e+00       4.246380e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.142200e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026160e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543890e+05  2.400000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  9.900000e+01        6.000000e+00       2.300000e+01   

       days_since_prior_order  add_to_cart_order   reordered     max_order  \
count              30328763.0         32404859.0  32404859.0  3.240486e+07   
mean                      NaN                NaN         NaN  3.305217e+01   
std                       0.0                0.0         0.0  2.515525e+01   
min                       0.0                1.0         0.0  1.000000e+00   
25%                       5.0                3.0         0.0  1.300000e+01   
50%                       8.0                6.0         1.0  2.600000e+01   
75%                      15.0               11.0         1.0  4.700000e+01   
max                      30.0              145.0         1.0  9.900000e+01   

          avg_price  median_frequency  
count  3.240486e+07        32404854.0  
mean   1.198023e+01               NaN  
std    8.324227e+01               0.0  
min    1.000000e+00               0.0  
25%    7.387298e+00               6.0  
50%    7.824786e+00               8.0  
75%    8.254023e+00              13.0  
max    2.500542e+04              30.0

# 06. Merging Data

In [38]:
# Merging new df_custom dataframe with old df_ords_prods_merge_agg_charts dataframe in a new df_ords_prods_customs dataframe
# Attempt a merge

df_ords_prods_customs = df_ords_prods_merge_agg_charts.merge(df_custom, on = 'User_ID', how = 'inner')

In [39]:
# Checking merged data

df_ords_prods_customs.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2         907        Premium Sliced Bacon       106             12    20.0   
3         907        Premium Sliced Bacon       106             12    20.0   
4        1000                    Apricots        18             10    12.9   

   order_id  User_ID  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2   3160996      138             1                   5                 13   
3   2254091      138            10                   5                 14   
4    505689      138             9                   6                 12   

   ...  order_frequency_flag  First_Name  Last_Name Gender      STATE Age  \
0  ...     Frequent customer     Charles        Cox   Male  Minnesota  81   
1  ...     Frequent customer     Charles        Cox   Male  Minnesota  81   
2  ...     Frequent customer     Charles        Cox   Male  Minnesota  81   
3  ...     Frequent customer     Charles        Cox   Male  Minnesota  81   
4  ...     Frequent customer     Charles        Cox   Male  Minnesota  81   

  Date_joined Number_of_dependants Family_status  Income  
0    8/1/2019                    1       married   49620  
1    8/1/2019                    1       married   49620  
2    8/1/2019                    1       married   49620  
3    8/1/2019                    1       married   49620  
4    8/1/2019                    1       married   49620  

[5 rows x 34 columns]

In [40]:
# Checking merged data

df_ords_prods_customs['_merge'].value_counts(dropna = False)

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

# 07. Checking merged data

In [41]:
# Renaming columns so that start with capitalised letter

df_ords_prods_customs.rename(columns = {'product_id' : 'Product_ID', 'product_name' : 'Product_name', 'aisle_id' : 'Aisle_ID', 'department_id' : 'Department_ID', 'prices' : 'Prices', 'order_id' : 'Order_ID', 'order_number' : 'Order_number', 'orders_day_of_week' : 'Orders_day_of_Week', 'order_hour_of_day' : 'Order_hour_of_Day', 'days_since_prior_order' : 'Days_since_prior_order', 'add_to_cart_order' : 'Add_to_cart_order', 'reordered' : 'Reordered', 'price_range_loc' : 'Price_range_loc', 'busiest_day' : 'Busiest_day', 'busiest_days' : 'Busiest_days', 'busiest_period_of_day' : 'Busiest_period_of_day', 'max_order' : 'Max_order', 'loyalty_flag' : 'Loyalty_flag', 'avg_price' : 'Avg_Price', 'spending_flag' : 'Spending_flag', 'median_frequency' : 'Median_frequency', 'order_frequency_flag' : 'Order_frequency_flag' }, inplace = True)

In [42]:
# Displaying the first 10 rows

df_ords_prods_customs.head(10)

Product_ID                    Product_name  Aisle_ID  Department_ID  \
0           1      Chocolate Sandwich Cookies        61             19   
1           1      Chocolate Sandwich Cookies        61             19   
2         907            Premium Sliced Bacon       106             12   
3         907            Premium Sliced Bacon       106             12   
4        1000                        Apricots        18             10   
5        3265   Mini Seedless Watermelon Pack        32              4   
6        4913            Table Water Crackers        78             19   
7        6184                     Clementines        32              4   
8        6901                     Raisin Bran       121             14   
9        7147  Classic Baby Creamers Potatoes        83              4   

   Prices  Order_ID  User_ID  Order_number  Orders_day_of_Week  \
0     5.8   3139998      138            28                   6   
1     5.8   1977647      138            30                   6   
2    20.0   3160996      138             1                   5   
3    20.0   2254091      138            10                   5   
4    12.9    505689      138             9                   6   
5     1.7    960220      138            17                   5   
6     4.4    894221      138            23                   5   
7     4.3   1977647      138            30                   6   
8     2.9   1154289      138            18                   1   
9     9.0    930730      138             2                   0   

   Order_hour_of_Day  ...  Order_frequency_flag  First_Name  Last_Name Gender  \
0                 11  ...     Frequent customer     Charles        Cox   Male   
1                 17  ...     Frequent customer     Charles        Cox   Male   
2                 13  ...     Frequent customer     Charles        Cox   Male   
3                 14  ...     Frequent customer     Charles        Cox   Male   
4                 12  ...     Frequent customer     Charles        Cox   Male   
5                 14  ...     Frequent customer     Charles        Cox   Male   
6                 13  ...     Frequent customer     Charles        Cox   Male   
7                 17  ...     Frequent customer     Charles        Cox   Male   
8                  7  ...     Frequent customer     Charles        Cox   Male   
9                 12  ...     Frequent customer     Charles        Cox   Male   

       STATE Age Date_joined Number_of_dependants Family_status  Income  
0  Minnesota  81    8/1/2019                    1       married   49620  
1  Minnesota  81    8/1/2019                    1       married   49620  
2  Minnesota  81    8/1/2019                    1       married   49620  
3  Minnesota  81    8/1/2019                    1       married   49620  
4  Minnesota  81    8/1/2019                    1       married   49620  
5  Minnesota  81    8/1/2019                    1       married   49620  
6  Minnesota  81    8/1/2019                    1       married   49620  
7  Minnesota  81    8/1/2019                    1       married   49620  
8  Minnesota  81    8/1/2019                    1       married   49620  
9  Minnesota  81    8/1/2019                    1       married   49620  

[10 rows x 34 columns]

In [43]:
# Checking the number of rows and columns in df_ords_prods_customs

df_ords_prods_customs.shape

(32404859, 34)

# 08. Export data

In [44]:
# Exporting data

df_ords_prods_customs.to_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_customers.pkl'))